### PR TITLE
Add the parser to parse time.

### PR DESCRIPTION
**Pull request by Shylock-Hg**

### CODE REVIEW

File: src/common/time/TimeParser.cpp [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+             return Status::Error("Unexpected token `%s'.", toString(n.type));
+         }
+     }},
+    {kDateDay,
+     [](Token t, Token n, DateTime& val, ExpectType type) -> StatusOr<State> {
+         if (t.type == TokenType::kDateDelimiter) {
+             return kDateDay;
+         }
+         if (t.type == TokenType::kNumber) {
+             if (t.val < 1 || t.val > 31) {
```
Comment (laura-ding) [inline src/common/time/TimeParser.cpp:87]: Inaccurate judgment, need to judge with the month.

Comment (Shylock-Hg) [reply src/common/time/TimeParser.cpp:87]: It's only limit checking, when create date will check the date validation.

**Review by darionyaphet — APPROVED** [review]

**Comment by monadbobo** [comment]
There are two problems:
1. does our current timestamp not support integers anymore?
2. if it's a string, which ISO standard are we parse according to?

**Comment by Shylock-Hg** [comment]
> There are two problems:
> 
> 1. does our current timestamp not support integers anymore?
> 2. if it's a string, which ISO standard are we parse according to?

1. The pr don't affect the timestamp
2. Follow the opencypher

File: src/common/time/TimeParser.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        TokenType type;
+        int32_t val;   // only used for number token
+        std::string str;
+    };
+
+    static std::string toString(const Token& t);
+
+    // TODO(shylock) support weeks/days format when support the convertion
+    // The state of current parser
+    enum State {
```
Comment (monadbobo) [inline src/common/time/TimeParser.h:85]: enum class.

File: src/common/time/TimeParser.cpp [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+                try {
+                    t.val = folly::to<int32_t>(digits);
+                } catch (std::exception &e) {
+                    return Status::Error("%s", e.what());
+                }
+                t.type = TokenType::kNumber;
+                tokens_.emplace_back(t);
+                digits.clear();
+            }
+        } else if (kTimeDelimiter == *c) {
```
Comment (monadbobo) [inline src/common/time/TimeParser.cpp:391]: Maybe it's more intuitive to use switch cases?

File: src/common/time/TimeParser.cpp [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+#include "common/time/TimezoneInfo.h"
+
+namespace nebula {
+namespace time {
+
+/*static*/ const std::vector<TimeParser::Component> TimeParser::dateTimeStates = {
+    // State, ShiftMap
+    {kInitial,
+     [](Token, Token, Context &ctx) -> StatusOr<State> {
+         if (ctx.type == ExpectType::kDateTime) {
```
Comment (monadbobo) [inline src/common/time/TimeParser.cpp:19]: Would it be better to use a hashmap here?

File: src/common/time/TimeParser.h (hunk 2 above)
Comment (Shylock-Hg) [reply src/common/time/TimeParser.h:85]: This used for array index.

File: src/common/time/TimeParser.cpp (hunk 3 above)
Comment (Shylock-Hg) [reply src/common/time/TimeParser.cpp:391]: There is a condition `isdigit`.

File: src/common/time/test/TimeParserTest.cpp [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+#include "common/base/Base.h"
+#include "common/time/TimeParser.h"
+
+TEST(TimeParser, DateTime) {
+    {
+        nebula::time::TimeParser parser;
+        auto result = parser.parseDateTime("2019-01-03T22:22:3.2333");
+        ASSERT_TRUE(result.ok()) << result.status();
+        EXPECT_EQ(nebula::DateTime(2019, 1, 3, 22, 22, 3, 2333), result.value());
```
[the remark this answers is not in the document]
Comment (Shylock-Hg) [reply src/common/time/test/TimeParserTest.cpp:16]: Yes it's just the utc time.

[the remark this answers is not in the document]
Comment (Shylock-Hg) [reply src/common/time/test/TimeParserTest.cpp]: Got it. The process of fraction of second had be corrected.

File: src/common/time/TimeParser.cpp [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+
+#include "common/time/TimeParser.h"
+#include <sstream>
+#include "common/time/TimeUtils.h"
+#include "common/time/TimezoneInfo.h"
+
+namespace nebula {
+namespace time {
+
+/*static*/ const std::vector<TimeParser::Component> TimeParser::dateTimeStates = {
```
Comment (yixinglu) [inline src/common/time/TimeParser.cpp:15]: please try to use FSM to solve these states transfer issue, that's powerful tool for string parser, need not any branches.

Comment (Shylock-Hg) [reply src/common/time/TimeParser.cpp:16]: I'm not sure about your meaning. Could you detail the structure about the FSM you means, and how to replace the branches.

File: src/common/datatypes/Date.cpp [hunk 7]
```diff
@@ -101,21 +106,25 @@ std::string Date::toString() const {
...
 std::string Time::toString() const {
+    auto microsecStr = folly::stringPrintf("%.9f", static_cast<uint32_t>(microsec) / 1000000.0);
+    auto decimalPart = decimal(microsecStr);
     // It's in current timezone already
-    return folly::stringPrintf("%02d:%02d:%02d.%06d", hour, minute, sec, microsec);
+    return folly::stringPrintf("%02d:%02d:%02d%s", hour, minute, sec, decimalPart.c_str());
 }
 
 std::string DateTime::toString() const {
+    auto microsecStr = folly::stringPrintf("%.9f", static_cast<uint32_t>(microsec) / 1000000.0);
```
Comment (Aiee) [inline src/common/datatypes/Date.cpp:116]: Why do we keep 9 digits for ms instead of 6? Nebula's time precision is different from openCypher.

Comment (Shylock-Hg) [reply src/common/datatypes/Date.cpp:116]: What's the difference? They are both fraction of second.

**Comment by Shylock-Hg** [comment]
Move to https://github.com/vesoft-inc/nebula/pull/2770

**Pull request closed without merging**